### PR TITLE
Refactor the project in accordance with the new Godot Android plugin v1 architecture

### LINKED ISSUE #93: Version 2.0
@NeoSpark314 @BastiaanOlij Now that the Android plugin re-architecture has been merged (https://github.com/godotengine/godot/pull/33682 and https://github.com/godotengine/godot/pull/36336), I'll be merging the [2.0 branch](https://github.com/GodotVR/godot_oculus_mobile/tree/2.0) into master.

Can you take a look at the 2.0 branch and ensure that the migration to the new workflow won't be an issue.

Thanks!

### PR DESCRIPTION
**Pull request by m4gr3d**
This transforms the original project into an Android library project, which has the nice side effect of simplifying integration with Android Studio and gradle.

Resolves #93

### CODE REVIEW

**Comment by m4gr3d** [comment]
@NeoSpark314 @BastiaanOlij ping!

**Comment by NeoSpark314** [comment]
I get the following error when trying to deploy the demo project (after I added the empty settings.gradle file):

image

do you have an idea what could cause this?

File: README.md [hunk 1]
```diff
@@ -23,68 +23,70 @@ Setup
...
+    - If deploying only on Oculus Quest, this must be set to `6DOF`
+    - If deploying on Oculus Go, or on Oculus Go and Oculus Quest, 
+    this must be set to `3DOF and 6DOF`
+  - `Hand Tracking`: This is only supported on the **Oculus Quest**
+    - Select `None` if your app doesn't need hand tracking
+    - Select `Optional` if your app can use hand tracking, but doesn't require it (i.e: also works with controllers).
+    - Select `Required` if your app only uses hand tracking.
+- `Custom Template`
+  - `Use Custom Build` must be enabled.
+  - `Plugins` list must contain `OVRMobile`. This enables the `OVRMobile` plugin for the project.
```
Comment (NeoSpark314) [inline README.md]: What is this `Plugins` list? I can't find it in the export settings.

Comment (m4gr3d) [reply README.md]: @NeoSpark314 To test this change, you must build the latest version of the Godot `3.2` branch and use that.

**Comment by m4gr3d** [comment]
> I get the following error when trying to deploy the demo project (after I added the empty settings.gradle file):
> 
> image
> 
> do you have an idea what could cause this?

Make sure you build both the `release` and `release_debug` Godot shared libraries via scons prior to building the Godot Android templates.
Once that's done, reinstall the Android custom build templates using the newly built templates.
The error you're seeing occurs when one of the target (`release` in this case) is missing from the generated Godot templates.

**Comment by NeoSpark314** [comment]
Thank you @m4gr3d; now I can export/deploy an .apk; I also updated to the recent 3.2 branch and now have the plugins setting in the export but I think I'm still missing something as the plugin is still not added to the generated .apk file it seems:

I have in demo/android/plugins/ the .aar files generated from the build (is there a difference between OVRMobile-debug/release.aar and plugin-debug/release.aar?)

And I have added `OVRMobile` to the Plugins parameter in the Custom Template export options.

But the resulting .apk does not include the scripts and binaries from the OVRMobile plugin.

Also is it correct that with new system you will not be able anymore to read the OvrVrApiTypes.gd script files form the ovrmobile plugin in the game project anymore?

**Comment by m4gr3d** [comment]
@NeoSpark314 Only one of the `OVRMobile*.aar` file should be added to the `demo/android/plugins` directory otherwise they'll collide with each other.

How do you check the resulting apk? A `aar` file is merged into the final apk so you won't see it as a standalone file if you look into the apk, although you can see its component if you look at the classes directory or assets directory.

All the existing functionality is conserved with the new system, nothing is lost :)

**Comment by NeoSpark314** [comment]
I removed all other .aar files now and have only the OVRMobile-Debug.aar in the plugins folder;
I am checking in the exported .apk for the plugin .gd files and also if the libvrapi.so and libgodot_ovrmobile.so file are in there. (When I start on the quest I get the error that the .gd script files are not found).

To me it looks like the .aar file is not merged at all (I'm probably still missing some setting or parameter somehwere):

This is the content of the lib folder in my exported .apk
image

This is the content of the OVRMobile-debug.aar file I have in the plugins folder (here the libraries are in a jni subfolder):
image

And these are my export settings (I export debug):
image

**Comment by m4gr3d** [comment]
@NeoSpark314 can you also provide a screenshot for the directory containing the AAR file, as well as a screenshot of the build.gradle file.

**Comment by NeoSpark314** [comment]
Here is my android/plugins folder:
image

and this is the build folder created by godot:
image

Here are also the .aar file and the generated build.gradle if this contains useful information.

[build.gradle.zip](https://github.com/GodotVR/godot_oculus_mobile/files/4403822/build.gradle.zip)
[OVRMobile-debug.aar.zip](https://github.com/GodotVR/godot_oculus_mobile/files/4403832/OVRMobile-debug.aar.zip)

**Comment by m4gr3d** [comment]
@NeoSpark314 either your 3.2 branch is not up-to-date, or the Godot templates file were not regenerated after updating it.
The `build.gradle` file you attached is the older version and doesn't match the [latest from the 3.2 branch](https://github.com/godotengine/godot/blob/3.2/platform/android/java/app/build.gradle#L52). 

Once you regenerate the Godot templates, make sure to update your build of Godot with the new files (location typically varies based on the OS), nuke your current `android` directory, and reinstall it from the Godot menu interface.

**Comment by NeoSpark314** [comment]
Ok; I cleaned and rebuild everything with the current 3.2 master; now I get the following error:
image
Do I need to downgrade my NDK or is this only a configuration issue?

**Comment by m4gr3d** [comment]
> Ok; I cleaned and rebuild everything with the current 3.2 master; now I get the following error:
> image
> Do I need to downgrade my NDK or is this only a configuration issue?

Someone on IRC was having the same issue and it was caused by environment variables; i.e: Android Studio and the shell used by Godot see two different things.
You should be able to solve the issue via this [stack overflow post](https://stackoverflow.com/questions/60404457/android-no-version-of-ndk-matched-the-requested-version).

**Comment by NeoSpark314** [comment]
@m4gr3d now it works; thanks for your help, patience and great work on restructuring the whole android build system!  Now I think I have seen enough of the new system to potentially help others too when they run into problems :-)

For now I solved the NDK verison system by adding the specific version to the android section in the build.gradle. Not sure where it picks up the older version number from (I have no android studio installed).

**Comment by m4gr3d** [comment]
@NeoSpark314 I should probably include steps to install a specific version of the NDK, and add that version to the gradle build file to avoid errors like the one you ran into.

**Comment by m4gr3d** [comment]
@NeoSpark314 Did you use version `21.0.6...` as the module ndk version?

**Comment by NeoSpark314** [comment]
@m4gr3d yes; I used
```
android {
    ndkVersion "21.0.6113669"
...
} 
```

**Comment by m4gr3d** [comment]
@NeoSpark314 I've updated the `README` and the gradle build file to specific the NDK version (version `21.0.6113669`)

**Pull request merged**